src/storage.h

```diff
@@ -0,0 +1,226 @@
+#ifndef CYCLUS_STORAGES_STORAGE_H_
+#define CYCLUS_STORAGES_STORAGE_H_
+
+#include <string>
+#include <list>
+#include <vector>
+
+#include "cyclus.h"
+
+// forward declaration
+namespace storage {
+class Storage;
+} // namespace storage
+
+
+namespace storage {
+/// @class Storage
+///
+/// This Facility is intended to hold materials for a user specified
+/// amount of time in order to model a storage facility with a certain
+/// residence time or holdup time.
+/// The Storage class inherits from the Facility class and is
+/// dynamically loaded by the Agent class when requested.
+///
+/// @section intro Introduction
+/// This Agent was initially developed to support the fco code-to-code 
+/// comparison.
+/// It's very similar to the "NullFacility" of years 
+/// past. Its purpose is to hold materials and release them only  
+/// after some period of delay time.
+///
+/// @section agentparams Agent Parameters
+/// in_commods is a vector of strings naming the commodities that this facility receives
+/// out_commods is a string naming the commodity that in_commod is stocks into
+/// residence_time is the minimum number of timesteps between receiving and offering
+/// in_recipe (optional) describes the incoming resource by recipe
+/// 
+/// @section optionalparams Optional Parameters
+/// max_inv_size is the maximum capacity of the inventory storage
+/// throughput is the maximum processing capacity per timestep
+///
+/// @section detailed Detailed Behavior
+/// 
+/// Tick:
+/// Nothing really happens on the tick. 
+///
+/// Tock:
+/// On the tock, any material that has been waiting for long enough (delay 
+/// time) is placed in the stocks buffer.
+///
+/// Any brand new inventory that was received in this timestep is placed into 
+/// the processing queue to begin waiting. 
+/// 
+/// Making Requests:
+/// This facility requests all of the in_commod that it can.
+///
+/// Receiving Resources:
+/// Anything of the in_commod that is received by this facility goes into the 
+/// inventory.
+///
+/// Making Offers:
+/// Any stocks material in the stocks buffer is offered to the market.
+///
+/// Sending Resources:
+/// Matched resources are sent immediately.
+class Storage 
+  : public cyclus::Facility,
+    public cyclus::toolkit::CommodityProducer {
+ public:  
+  /// @param ctx the cyclus context for access to simulation-wide parameters
+  Storage(cyclus::Context* ctx);
+  
+  #pragma cyclus decl
+
+  #pragma cyclus note {"doc": "A storage facility converts from one " \
+                              "commodity to another, with an optional delay."}
```

Comment:
This needs a good, detailed description of how the storage facility operates generally.  As an example, here is what is in the cycamore::Reactor's note doc string:

```
#pragma cyclus note { \
"doc": \
  "Reactor is a simple, general reactor based on static compositional" \
  " transformations to model fuel burnup.  The user specifies a set of input" \
  " fuels and corresponding burnt compositions that fuel is transformed to when" \
  " it is discharged from the core.  No incremental transmutation takes place." \
  " Rather, at the end of an operational cycle, the batch being discharged from" \
  " the core is instantaneously transmuted from its original fresh fuel" \
  " composition into its spent fuel form." \
  "\n\n" \
  "Each fuel is identified by a specific input commodity and has an associated" \
  " input recipe (nuclide composition), output recipe, output commidity, and" \
  " preference.  The preference identifies which input fuels are preferred when" \
  " requesting.  Changes in these preferences can be specified as a function of" \
  " time using the pref_change variables.  Changes in the input-output recipe" \
  " compositions can also be specified as a function of time using the" \
  " recipe_change variables." \
  "\n\n" \
  "The reactor treats fuel as individual assemblies that are never split," \
  " combined or otherwise treated in any non-discrete way.  Fuel is requested" \
  " in full-or-nothing assembly sized quanta.  If real-world assembly modeling" \
  " is unnecessary, parameters can be adjusted (e.g. n_assem_core, assem_size," \
  " n_assem_batch).  At the end of every cycle, a full batch is discharged from" \
  " the core consisting of n_assem_batch assemblies of assem_size kg. The" \
  " reactor also has a specifiable refueling time period following the end of" \
  " each cycle at the end of which it will resume operation on the next cycle" \
  " *if* it has enough fuel for a full core; otherwise it waits until it has" \
  " enough fresh fuel assemblies." \
  "\n\n" \
  "In addition to its core, the reactor has an on-hand fresh fuel inventory" \
  " and a spent fuel inventory whose capacities are specified by n_assem_fresh" \
  " and n_assem_spent respectively.  Each time step the reactor will attempt to" \
  " acquire enough fresh fuel to fill its fresh fuel inventory (and its core if" \
  " the core isn't currently full).  If the fresh fuel inventory has zero" \
  " capacity, fuel will be ordered just-in-time after the end of each" \
  " operational cycle before the next begins.  If the spent fuel inventory" \
  " becomes full, the reactor will halt operation at the end of the next cycle" \
  " until there is more room.  Each time step, the reactor will try to trade" \
  " away as much of its spent fuel inventory as possible." \
  "\n\n" \
  "When the reactor reaches the end of its lifetime, it will discharge all" \
  " material from its core and trade away all its spent fuel as quickly as" \
  " possible.  Full decommissioning will be delayed until all spent fuel is" \
  " gone.  If the reactor has a full core when it is decommissioned (i.e. is" \
  " mid-cycle) when the reactor is decommissioned, half (rounded up to nearest" \
  " int) of its assemblies are transmuted to their respective burnt" \
  " compositions." \
  "", \
}
```